src/de/packet_parser.rs

```diff
@@ -0,0 +1,133 @@
+use crate::{warn, MessageType, Property, ProtocolError as Error};
+use bit_field::BitField;
+use heapless::Vec;
+
+pub(crate) struct PacketParser<'a> {
+    pub buffer: &'a [u8],
+    index: core::cell::RefCell<usize>,
```

Comment:
I wonder what the original reason for the `RefCell` was, whether we can get rid of it now, and instead make the methods require `&mut self`.

Reply:
The reason is because we are borrowing data out of `buffer` to create `Property` objects. As a result of the immutable borrow for the `buffer`, `self` can no longer mutate. Hence, why `index` is stored in a `RefCell`.

I tried removing it during this refactor, but discovered that issue.

Reply:
Ack